auth/src/main/java/org/aerogear/mobile/auth/authenticator/OIDCAuthenticatorImpl.java

```diff
@@ -0,0 +1,208 @@
+import org.aerogear.mobile.core.executor.AppExecutors;
+import org.aerogear.mobile.core.http.HttpRequest;
+import org.aerogear.mobile.core.http.HttpServiceModule;
+import org.aerogear.mobile.core.http.OkHttpServiceModule;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.Principal;
+
+/**
+ * Authenticates the user by using OpenID Connect.
+ */
+public class OIDCAuthenticatorImpl extends AbstractAuthenticator {
+
+    private AuthState authState;
+
+    private AuthorizationService authService;
+
+    private final KeycloakConfiguration keycloakConfiguration;
+    private final AuthServiceConfiguration authServiceConfiguration;
+
+    private final Context appContext;
+
+    private Callback authCallback;
+
+    private final AuthStateManager authStateManager;
+
+    /**
+     * Creates a new OIDCAuthenticatorImpl object
+     *
+     * @param serviceConfiguration {@link ServiceConfiguration}
+     * @param authServiceConfiguration {@link AuthServiceConfiguration}
+     */
+    public OIDCAuthenticatorImpl(final ServiceConfiguration serviceConfiguration, final AuthServiceConfiguration authServiceConfiguration, final Context context, final AuthStateManager authStateManager) {
+        super(serviceConfiguration);
+        this.keycloakConfiguration = new KeycloakConfiguration(serviceConfiguration);
+        this.authServiceConfiguration = authServiceConfiguration;
+        this.appContext = context;
+        this.authStateManager = authStateManager;
+    }
+
+    /**
+     * Builds a new OIDCUserPrincipalImpl object after the user's credential has been authenticated
+     *
+     * @param authOptions the OIDC authentication options
+     * @return a new OIDCUserPrincipalImpl object with the user's identity that was decoded from the user's credential
+     * @throws AuthenticationException
+     */
+    @Override
+    public void authenticate(final AuthenticateOptions authOptions, final Callback<Principal> callback) {
+        this.authCallback = callback;
+        OIDCAuthenticateOptions oidcAuthenticateOptions = (OIDCAuthenticateOptions) (authOptions);
+        performAuthRequest(oidcAuthenticateOptions.getFromActivity(), oidcAuthenticateOptions.getResultCode());
+    }
+
+    // Authentication code
+    private void performAuthRequest(final Activity fromActivity, final int resultCode) {
+        AuthorizationServiceConfiguration authServiceConfig = new AuthorizationServiceConfiguration(
+            this.keycloakConfiguration.getAuthenticationEndpoint(),
+            this.keycloakConfiguration.getTokenEndpoint()
+        );
+        this.authState = new AuthState(authServiceConfig);
+
+        AppAuthConfiguration.Builder appAuthConfigurationBuilder = new AppAuthConfiguration.Builder()
+            .setBrowserMatcher(new BrowserBlacklist(
+                VersionedBrowserMatcher.CHROME_CUSTOM_TAB));
+
+        if (this.authServiceConfiguration.isAllowSelfSignedCertificate()) {
```

Comment:
Is this check aimed at the tests only? If so, maybe make a note to avoid confusion?

Reply:
I will address this in a follow up PR